File: Bronze/2018-01/lifeguards_jan2018.java
// Lifeguards - USACO Bronze January 2018 (http://www.usaco.org/index.php?page=viewproblem2&cpid=784)
// This problem was completed on September 27, 2020, in 35 minutes, with all 10/10 test cases passed (second try)

import java.io.File;
import java.io.FileWriter;
import java.io.IOException;
import java.io.PrintWriter;
import java.util.Arrays;
import java.util.Scanner;

public class lifeguards_jan2018 {
    static int numLifeguards;
    static int[][] lifeguardShifts;

    public static int lifeguards() {
        int maxLifeguardTime = 0;

        for (int i=0; i<numLifeguards; i++) { // i is the index for which lifeguard to fire
            int currLifeguardTime = 0;

            // first generate a timeline
            int[] timeline = new int[numLifeguards*2-2];
            for (int j=0, idx=0; j<numLifeguards; j++) {
                if (j == i) {
                    continue;
                }
                timeline[idx++] = lifeguardShifts[j][0];
                timeline[idx++] = lifeguardShifts[j][1];
            }
            Arrays.sort(timeline);

            // iterate through the timeline: add current lifeguard time, update lifeguard num
            int currLifeguardCnt = 0;

            for (int j=0; j<timeline.length; j++) {
                if (currLifeguardCnt > 0) {
                    currLifeguardTime += (timeline[j] - timeline[j-1]);
                }

                // update lifeguard count
                if (isShiftStartTime(timeline[j])) {
                    currLifeguardCnt++;
                } else {
                    currLifeguardCnt--;
                }
            }

            maxLifeguardTime = Math.max(maxLifeguardTime, currLifeguardTime);
        }

        return maxLifeguardTime;
    }

    public static boolean isShiftStartTime(int time) {
        for (int i=0; i<numLifeguards; i++) {
            if (lifeguardShifts[i][0] == time) { // check for start time
                return true;
            } else if (lifeguardShifts[i][1] == time) { // check for end time
                return false;
            }
        }

        return true; // this should never happen, avoid compilation errors
    }

    public static void main(String[] args) throws IOException {
        // input
        String problemName = "lifeguards";
        Scanner sc = new Scanner(new File(problemName + ".in"));

        numLifeguards = sc.nextInt();
        lifeguardShifts = new int[numLifeguards][2];
        for (int i = 0; i< numLifeguards; i++) {
            lifeguardShifts[i][0] = sc.nextInt();
            lifeguardShifts[i][1] = sc.nextInt();
        }

        // algorithm
        int numLifeguardTime = lifeguards();

        // output
        PrintWriter out = new PrintWriter(new FileWriter(problemName + ".out"));
        out.println(numLifeguardTime);
        out.close();
    }
}

File: Bronze/2018-01/lifeguards_jan2018.cpp
#include <bits/stdc++.h>

using namespace std;

int n;
int lifeguards[105][2];

int main() {
    ifstream fin("lifeguards.in");
    ofstream fout("lifeguards.out");

    fin >> n;
    for (int i=1; i<=n; i++) fin >> lifeguards[i][0] >> lifeguards[i][1];

    pair<int, int> timeline[2*n];

    int result = 0;

    for (int i=1; i<=n; i++) {
        for (int j=1, idx=1; j<=n; j++) {
            if (i == j) continue;

            timeline[idx++] = make_pair(lifeguards[j][0], 1);
            timeline[idx++] = make_pair(lifeguards[j][1], -1);
        }

        sort(timeline+1, timeline+2*n-1);

        int guardCnt = 0;
        int time = 0;

        for (int j=1; j<=2*n-2; j++) {
            if (guardCnt > 0)
                time += (timeline[j].first - timeline[j-1].first);
            
            guardCnt += timeline[j].second;
        }

        result = max(result, time);
    }

    fout << result << endl;
}